### PR TITLE
Support permissionless price feed updates

### PR DESCRIPTION
**Pull request by nickguo**
This feature saves operational costs for the program authority.

Rather than the authority having to post frequent on chain price oracle updates, the user can post the oracle price update at the time of their transaction.

The authority hosts a backend service that the user can query to obtain a fresh price update payload signed by the authority. The user then includes this payload in their transaction.

Compute cost for just `SetCustomOraclePricePermissionless` is 22k.  The ed25519 program cost is not showing up in the logs for some reason, but overall budget is showing as 400k instead of 200k.
Screenshot 2023-10-13 at 7 49 32 AM

### CODE REVIEW

File: programs/perpetuals/tests/anchor/basic.ts [hunk 1]
```diff
@@ -124,7 +125,7 @@ describe("perpetuals", () => {
     expect(JSON.stringify(pool)).to.equal(JSON.stringify(poolExpected));
 
     await tc.removePool();
-    tc.ensureFails(tc.program.account.pool.fetch(tc.pool.publicKey));
+    await tc.ensureFails(tc.program.account.pool.fetch(tc.pool.publicKey));
```
Comment (nickguo) [inline programs/perpetuals/tests/anchor/basic.ts:128]: unrelated change;  `await` is needed for ensureFails.

File: programs/perpetuals/tests/native/utils/test_setup.rs [hunk 2]
```diff
@@ -205,14 +205,10 @@ impl TestSetup {
 
         // Initialize users token accounts for each mints
         {
-            let mints_pubkeys: Vec<Pubkey> =
-                mints.values().into_iter().map(|info| info.pubkey).collect();
+            let mints_pubkeys: Vec<Pubkey> = mints.values().map(|info| info.pubkey).collect();
```
Comment (nickguo) [inline programs/perpetuals/tests/native/utils/test_setup.rs:208]: unrelated clippy format update

File: programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        // Custom oracle must first be initialized by authority before permissionless updates.
+        mut,
+        seeds = [b"oracle_account",
+                 pool.key().as_ref(),
+                 custody.mint.as_ref()],
+        bump
+    )]
+    pub oracle_account: Box<Account<'info, CustomOracle>>,
+
+    system_program: Program<'info, System>,
```
Comment (askibin) [inline programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs:47]: why do you need system program here?

Comment (nickguo) [reply programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs:47]: ah my bad, I started from a copy of `set_custom_oracle_price.rs` but don't need the system program since `init_if_needed` is no longer used.

removed! 👍

File: programs/perpetuals/src/instructions/set_permissionless_oracle_pubkey.rs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    #[account(
+        mut,
+        seeds = [b"custody",
+                 pool.key().as_ref(),
+                 custody.mint.as_ref()],
+        bump = custody.bump
+    )]
+    pub custody: Box<Account<'info, Custody>>,
+
+    system_program: Program<'info, System>,
```
Comment (nickguo) [inline programs/perpetuals/src/instructions/set_permissionless_oracle_pubkey.rs:47]: similar to above, removed this `system_program`. https://github.com/solana-labs/perpetuals/pull/33/files#r1357291887

File: programs/perpetuals/tests/anchor/cryptography.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import * as sha3 from "sha3";
+import { randomBytes } from "crypto";
+import * as secp256k1 from "secp256k1";
+
+export function generateSecp256k1Keypair() {
```
Comment (harshasomisetty) [inline programs/perpetuals/tests/anchor/cryptography.ts:5]: why do you need to generate a keypair yourself?

there is a Keypair object in the web3.js library, unless you have a reason for making your own?

File: programs/perpetuals/tests/anchor/cryptography.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    do {
+      privateKey = randomBytes(32)
+    } while (!secp256k1.privateKeyVerify(privateKey));
+    // Get the public key in uncompressed format.
+    const publicKey = secp256k1.publicKeyCreate(privateKey, false);
+    // The library always returns a 65 byte public key that starts with 0x04, so omit the first byte.
+    return [privateKey, publicKey.slice(1)];
+}
+
+export function secp256k1Sign(privateKey, message) {
```
Comment (harshasomisetty) [inline programs/perpetuals/tests/anchor/cryptography.ts:17]: can also straight sign with the anchor namespace .signer() command or along those lines

File: programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs (hunk 3 above)
Comment (harshasomisetty) [reply programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs:47]: idt you pushed this change

**Review by harshasomisetty — APPROVED** [review]

File: programs/perpetuals/tests/anchor/cryptography.ts (hunk 5 above)
Comment (nickguo) [reply programs/perpetuals/tests/anchor/cryptography.ts:5]: hahah yes good call;  we had initially set the signing scheme to be secp256k1 and I didn't see support for that in web3.js -- but we have since changed the scheme to ed25519!   removed secp256k1 utils

File: programs/perpetuals/tests/anchor/cryptography.ts (hunk 6 above)
Comment (nickguo) [reply programs/perpetuals/tests/anchor/cryptography.ts]: it seems those were for transactions;  went with TweetNaCl per the cookbook https://solanacookbook.com/references/keypairs-and-wallets.html#how-to-sign-and-verify-messages-with-wallets  please let me know if there is a better approach!

File: programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        mut,
+        seeds = [b"oracle_account",
+                 pool.key().as_ref(),
+                 custody.mint.as_ref()],
+        bump
+    )]
+    pub oracle_account: Box<Account<'info, CustomOracle>>,
+
+    /// CHECK: Needed for ed25519 signature verification, to inspect all instructions in this transaction.
+    #[account(address = IX_ID)]
```
Comment (askibin) [inline programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs:53]: could you keep it as `sysvar::instructions::ID`, would be easier to understand where IX_ID comes from. same for ED25519_ID.

File: programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+fn validate_ed25519_signature_instruction(
+    signature_ix: &Instruction,
+    expected_pubkey: &Pubkey,
+    custody_key: &Pubkey,
+    serialized_oracle: &[u8],
+) -> Result<()> {
+    require_eq!(
+        signature_ix.program_id,
+        ED25519_ID,
+        PerpetualsError::PermissionlessOraclePriceSignatureFailure
```
Comment (askibin) [inline programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs:102]: could you create additional error codes so it is easier to determine what went wrong?

File: programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    );
+    require_eq!(
+        signature_ix.accounts.len(),
+        0,
+        PerpetualsError::PermissionlessOraclePriceSignatureFailure
+    );
+
+    let verified_pubkey = &signature_ix.data[16..16 + 32];
+    let verified_message = &signature_ix.data[112..];
```
Comment (askibin) [inline programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs:111]: all this math is unclear, could you add a comment and explicitly check for data size before slicing

File: programs/perpetuals/src/state/oracle.rs [hunk 10]
```diff
@@ -49,6 +50,24 @@ pub struct CustomOracle {
...
+
+    pub fn set(&mut self, price: u64, expo: i32, conf: u64, ema: u64, publish_time: i64) {
+        self.price = price;
+        self.expo = expo;
+        self.conf = conf;
+        self.ema = ema;
+        self.publish_time = publish_time;
+    }
+
+    pub fn serialize(&self) -> [u8; 36] {
```
Comment (askibin) [inline programs/perpetuals/src/state/oracle.rs:62]: anchor uses borsch to serialize. by adding AnchorSerialize to the struct, you can serialize without writing this method

File: programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    pub publish_time: i64,
+}
+
+pub fn set_custom_oracle_price_permissionless<'info>(
+    ctx: Context<'_, '_, '_, 'info, SetCustomOraclePricePermissionless<'info>>,
+    params: &SetCustomOraclePricePermissionlessParams,
+) -> Result<()> {
+    // If the incoming publish time is behind or equal to the current, then no-op.
+    if params.publish_time <= ctx.accounts.oracle_account.publish_time {
+        msg!("Publish time is behind the current, so CustomOracle price did not update.");
```
Comment (askibin) [inline programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs:72]: This log message doesn't sound well. Lets reword it or remove entirely

File: programs/perpetuals/tests/anchor/test_client.ts [hunk 12]
```diff
@@ -706,6 +707,93 @@ export class TestClient {
...
+    {
+      custody_account: Pubkey,
+      price: u64,
+      expo: i32,
+      conf: u64,
+      ema: u64,
+      publish_time: i64,
+    }
+    */
+    let buffer = Buffer.alloc(68);
```
Comment (askibin) [inline programs/perpetuals/tests/anchor/test_client.ts:738]: you should be able to do it with anchor

File: programs/perpetuals/tests/anchor/cryptography.ts [hunk 13]
```diff
@@ -0,0 +1,6 @@
+import { Keypair } from "@solana/web3.js";
```
Comment (askibin) [inline programs/perpetuals/tests/anchor/cryptography.ts:1]: why do we need a separate file just for a single line? keep it simple

File: programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs (hunk 7 above)
Comment (nickguo) [reply programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs:53]: yes agreed!  updated

File: programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+pub struct SetCustomOraclePricePermissionlessParams {
+    pub price: u64,
+    pub expo: i32,
+    pub conf: u64,
+    pub ema: u64,
+    pub publish_time: i64,
+}
+
+pub fn set_custom_oracle_price_permissionless<'info>(
+    ctx: Context<'_, '_, '_, 'info, SetCustomOraclePricePermissionless<'info>>,
```
[the remark this answers is not in the document]
Comment (nickguo) [reply programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs:67]: not needed, removed.  the other instructions follow this format as well, should I make a separate PR to remove their explicit lifetimes?

File: programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs (hunk 11 above)
Comment (nickguo) [reply programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs:71]: ahh yeah sorry about that!  how is the updated message?

File: programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs (hunk 8 above)
Comment (nickguo) [reply programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs]: done!  made different error codes

File: programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs (hunk 9 above)
Comment (nickguo) [reply programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs]: yes!  added more checks above for the message data size, and also added comment for where the offsets came from

File: programs/perpetuals/src/state/oracle.rs (hunk 10 above)
Comment (nickguo) [reply programs/perpetuals/src/state/oracle.rs]: ahh this was fun!  previously I had rolled the encoding format by just stacking the fields.

Updated contract-side and client-side to use `borsh`.  It also cleaned up the code on both ends 👍

File: programs/perpetuals/tests/anchor/test_client.ts (hunk 12 above)
Comment (nickguo) [reply programs/perpetuals/tests/anchor/test_client.ts]: done with `borsh`

File: programs/perpetuals/tests/anchor/cryptography.ts (hunk 13 above)
Comment (nickguo) [reply programs/perpetuals/tests/anchor/cryptography.ts:1]: done, removed and replaced caller with the one liner `nacl.sign.detached`

File: programs/perpetuals/tests/anchor/test_client.ts [hunk 15]
```diff
@@ -706,6 +708,92 @@ export class TestClient {
...
+          signature: signature,
+        })
+      );
+    }
+    tx.add(
+      this.program.instruction.setCustomOraclePricePermissionless(
+        setCustomOraclePricePermissionlessParams,
+        {
+          accounts: {
+            user: user.publicKey,
```
Comment (askibin) [inline programs/perpetuals/tests/anchor/test_client.ts:770]: you don't need to explicitly specify unused default signer as anchor signs with provider.wallet anyway

File: programs/perpetuals/src/state/multisig.rs [hunk 16]
```diff
@@ -34,6 +34,7 @@ pub enum AdminInstruction {
     WithdrawFees,
     WithdrawSolFees,
     SetCustomOraclePrice,
+    SetPermissionlessOraclePubkey,
```
Comment (askibin) [inline programs/perpetuals/src/state/multisig.rs:37]: it doesn't need to be here, this enum only for admin instructions

File: programs/perpetuals/tests/anchor/basic.ts [hunk 17]
```diff
@@ -408,6 +419,84 @@ describe("perpetuals", () => {
...
+      tc.setCustomOraclePricePermissionless(
+        permissionlessOracleKeypair,
+        100,
+        tc.custodies[1],
+        null,
+        true
+      )
+    );
+
+    // Sending the permissionless update without malformed message should fail.
```
Comment (askibin) [inline programs/perpetuals/tests/anchor/basic.ts:485]: could you also add a test that sends a bogus signature?

File: programs/perpetuals/src/state/multisig.rs (hunk 16 above)
Comment (nickguo) [reply programs/perpetuals/src/state/multisig.rs:37]: ah my bad, remnant from the earlier standalone set-key instr!  removed

File: programs/perpetuals/tests/anchor/basic.ts (hunk 17 above)
Comment (nickguo) [reply programs/perpetuals/tests/anchor/basic.ts]: done! added above and renamed to `bogusKeypair`

File: programs/perpetuals/tests/anchor/test_client.ts (hunk 15 above)
Comment (nickguo) [reply programs/perpetuals/tests/anchor/test_client.ts]: that is good to know!!  removed `user` from the instruction and also dropped the signer here

File: programs/perpetuals/tests/anchor/test_client.ts [hunk 18]
```diff
@@ -706,6 +708,88 @@ export class TestClient {
...
+            custody: custody.custody,
+            oracleAccount: custody.oracleAccount,
+            systemProgram: SystemProgram.programId,
+            ixSysvar: anchor.web3.SYSVAR_INSTRUCTIONS_PUBKEY,
+          },
+        }
+      )
+    );
+
+    try {
```
Comment (askibin) [inline programs/perpetuals/tests/anchor/test_client.ts:779]: instead of building a custom instruction you could use `.preInstructions` , i.e. `this.program.methods.xxx(...).accounts(...)..preInstructions(your_ed2519_ix).rpc()`

File: programs/perpetuals/tests/anchor/test_client.ts [hunk 19]
```diff
@@ -706,6 +708,88 @@ export class TestClient {
...
 
+  setCustomOraclePricePermissionless = async (
+    permissionlessOracleKeypair: Keypair,
+    price: number,
+    custody,
+    publishTime?,
+    noSignatureVerification?,
+    messageOverwrite?
+  ) => {
+    const schema = {
```
Comment (askibin) [inline programs/perpetuals/tests/anchor/test_client.ts:719]: are you sure anchor doesn't let you serialize directly?

File: programs/perpetuals/tests/anchor/test_client.ts (hunk 18 above)
Comment (nickguo) [reply programs/perpetuals/tests/anchor/test_client.ts:764]: yes done!  built tx with `.preInstructions`

File: programs/perpetuals/tests/anchor/test_client.ts (hunk 19 above)
Comment (nickguo) [reply programs/perpetuals/tests/anchor/test_client.ts]: thanks for the push on this Alexey!  after some digging, yes we have access to the borsh serde that anchor uses.
```
this.program._coder.types.encode(
            "SetCustomOraclePricePermissionlessParams",
            setCustomOraclePricePermissionlessParams
          );
```
Updated to use the Anchor serializer instead, and removed the `borsh` .js dependency 👍

File: programs/perpetuals/tests/anchor/basic.ts [hunk 20]
```diff
@@ -408,6 +419,80 @@ describe("perpetuals", () => {
...
+      tc.getTime() - 20
+    );
+    oracle = await tc.program.account.customOracle.fetch(
+      tc.custodies[0].oracleAccount
+    );
+    // Oracle's value is still 500 instead of the attempted 400.
+    expect(JSON.stringify(oracle)).to.equal(JSON.stringify(oracleExpected));
+
+    // after test, set price back to the expected for other test cases.
+    await tc.setCustomOraclePricePermissionless(
```
Comment (askibin) [inline programs/perpetuals/tests/anchor/basic.ts:455]: could you add one more test please, I want to make sure setting priority fees and CUs still work. So please build a tx consisting of the following ixs: 1. ed25519 program 2. setComputeUnitLimit 3. SetComputeUnitPrice 4. oracle price update
https://solanacookbook.com/references/basic-transactions.html#how-to-change-compute-budget-fee-priority-for-a-transaction

Comment (nickguo) [reply programs/perpetuals/tests/anchor/basic.ts:477]: yes!  added a test case for `increaseComputeLimits` above.  confirmed tests pass as expected

File: programs/perpetuals/tests/anchor/test_client.ts (hunk 18 above)
Comment (askibin) [reply programs/perpetuals/tests/anchor/test_client.ts:775]: could you confirm that `preInstructions()` call adds instructions to the tx and not replaces the ones that were stored there before? Otherwise you would replace the call to Ed25519 program with set compute budget

Comment (nickguo) [reply programs/perpetuals/tests/anchor/test_client.ts:775]: Yes, added a print statement and saw that the successive `tx = tx.preInstructions(...)` calls add to each other
Screenshot 2023-10-14 at 5 27 19 PM

Also have a test case where the ed25519 signature is missing -- confirms that the tx fails with `PermissionlessOracleMissingSignature`.

File: programs/perpetuals/src/state/oracle.rs [hunk 21]
```diff
@@ -33,6 +33,7 @@ pub struct OraclePrice {
 pub struct OracleParams {
     pub oracle_account: Pubkey,
     pub oracle_type: OracleType,
+    pub permissionless_update_pubkey: Pubkey,
```
Comment (askibin) [inline programs/perpetuals/src/state/oracle.rs:36]: lets rename this field, it is actually a public key of the oracle authority that will sign the message on a backend. A better name would be `oracle_authority`

File: programs/perpetuals/tests/anchor/basic.ts [hunk 22]
```diff
@@ -20,6 +25,7 @@ describe("perpetuals", () => {
   let multisigExpected;
   let tokenExpected;
   let positionExpected;
+  let permissionlessOracleUpdateKey;
```
Comment (askibin) [inline programs/perpetuals/tests/anchor/basic.ts:28]: also lets rename it here and move to test_client.ts where all similar keys are generated

File: programs/perpetuals/src/state/oracle.rs (hunk 21 above)
Comment (nickguo) [reply programs/perpetuals/src/state/oracle.rs]: renamed 👍

File: programs/perpetuals/tests/anchor/basic.ts (hunk 22 above)
Comment (nickguo) [reply programs/perpetuals/tests/anchor/basic.ts]: done!

**Review by askibin — APPROVED** [review]
